In [1]:
import numpy as np

In [ ]:
# membuat matriks koefisien 
a = np.array([[2, 2, 1],
             [1, -1, 1],
             [3, 2, -2]], dtype=float)

# membuat vektor hasil
b = np.array([3,4,2], dtype=float)
# jumlah persamaan 
n = len(b)

# menampilkan matriks augmentasi [a | b]
print("Augmentasi Matriks: \n", np.column_stack((a, b)))

# eliminasi Gauss untuk membentuk matriks segitiga atas
# loop tiap kolom
for k in range(n):
    # cek pivot (diagonal utama matriks) apakah 0? 
    if a[k,k] == 0:
        # kalau iya, coba tukar baris
        for i in range(k, len(b)):      # cek baris di bawah
            if a[i, k] != 0:            # cari yang tidak nol         
                a[[k, i]] = a[[i, k]]   # tukar baris k (yang nol) dgn baris i (yang tdk nol) utk matriks a
                b[[k, i]] = b[[i, k]]   # tukar baris k (yang nol) dgn baris i (yang tdk nol) utk matriks b
                break

    # ambil pivot (diagonal utama)           
    pivot = a [k, k]

    # jika setelah swap pivot masih 0 berarti matriks tidak bisa diselesaikan
    if pivot == 0:  
        raise ValueError("Sistem tidak memiliki solusi unik atau tak terdefinisi.")
    
    # ubah semua pivot jadi 1 dengan membagi seluruh baris
    a[k,:] /= pivot
    b[k] /= pivot
    
    # buat semua elemen di bawah piot jadi 0
    for i in range(k+1, n):
        factor = a[i, k] 
        a[i, k:] -= factor * a[k, k:]
        b[i] -= factor * b[k]

# hasil setelah eliminasi
print("\nMatriks setelah eliminasi Gauss:")
print(np.column_stack((a, b)))

# substitusi mundur untuk mendapatkan hasil (cari nilai x dari bawah ke atas)
print("\nmatriks a")
print(a)
print("\nmatriks b")
print(b)

# buat array solusi 
x = np.zeros(n)

# mulai dari variabel terakhir (z)
x[-1] = b[-1] / a[-1, -1]

# loop dari bawah ke atas
for i in range(n-2, -1, -1):
    # hitung jumlah a[i][j] * x[j] untuk j di kanan diagonal
    sum_val = b[i] - np.dot(a[i, i+1:], x[i+1:])
    # cari nilai x[i]
    x[i] = sum_val / a[i, i]

print("\nSolusi:")
for i in range(len(x)):
    print("x%d = %.4f" % (i+1, x[i]))

Augmentasi Matriks: 
 [[ 2.  2.  1.  3.]
 [ 1. -1.  1.  4.]
 [ 3.  2. -2.  2.]]

Matriks setelah eliminasi Gauss:
[[ 1.    1.    0.5   1.5 ]
 [-0.    1.   -0.25 -1.25]
 [-0.   -0.    1.    1.  ]]

matriks a
[[ 1.    1.    0.5 ]
 [-0.    1.   -0.25]
 [-0.   -0.    1.  ]]

matriks b
[ 1.5  -1.25  1.  ]

Solusi:
x1 = 2.0000
x2 = -1.0000
x3 = 1.0000


### CATATAN #1

Matriks `a` adalah matriks koefisien dari sistem persamaan linear.

```python
a = np.array([[2, 2, 1],
              [1, -1, 1],
              [3, 2, -2]], dtype=float)
```

Vektor `b` adalah hasil dari tiap persamaan (ruas kanan).

```python
b = np.array([3, 4, 2], dtype=float)
```

MAKA SPLNYA:
$$
\begin{aligned}
&2x + 2y + z = 3 \\
&x - y + z = 4 \\
&3x + 2y - 2z = 2
\end{aligned}
$$

Karena `b = [3, 4, 2]`, berarti `n = 3`  
Artinya ada 3 persamaan dan 3 variabel (x, y, z)

```python
print("Augmentasi Matriks: \n", np.column_stack((a, b)))
```

`np.column_stack` adalah fungsi dari NumPy yang digunakan untuk menggabungkan array secara horizontal (kolom)

Jadi karena `a` = matriks (2D) dan `b` = vektor (1D)
maka hasilnya `b` ditaruh sebagai kolom tambahan di sebelah kanan `a`

HASILNYA

$$
\left[
\begin{array}{ccc|c}
2 & 2 & 1 & 3 \\
1 & -1 & 1 & 4 \\
3 & 2 & -2 & 2
\end{array}
\right]
$$

---


### CATATAN #2

Eliminasi Gauss → mengubah matriks augmentasi menjadi matriks segitiga atas (semua elemen di bawah diagonal utama menjadi 0).

```python
for k in range(n):
```
- `n` = 3
- `k` = 0 → kolom pertama
- `k` = 1 → kolom kedua
- `k` = 2 → kolom ketiga


```python
if a[k,k] == 0:
    for i in range(k, len(b)):
        if a[i, k] != 0:
            a[[k, i]] = a[[i, k]]
            b[[k, i]] = b[[i, k]]
            break
```

``` a[k,k] ``` itu kita jadiin pivot, dimana pivot kita berarti diagonal utama, jadi kalau kita pakai contoh matriks `a` tadi, berarti diagonal utamanya 2, -1, 2

Contoh
- `k` = 0 → `a[0,0]` = 2
- `k` = 1 → `a[1,1]` = -1
- `k` = 2 → `a[2,2]` = -2

Kalau pivot kita 0 `if a[k,k] == 0:` kita tidak bisa melanjutkan eliminasi gauss, maka dari itu kita buat kondisi kalau pivotnya 0 maka bagian code yang jalan adalah
```python
    for i in range(k, len(b)):
        if a[i, k] != 0:
            a[[k, i]] = a[[i, k]]
            b[[k, i]] = b[[i, k]]
            break
```

Artinya
1. Cari baris di bawah
2. Cari yang kolomnya tidak nol
3. Tukar baris (swap)

---

### CATATAN #3

CONTOH VISUALISASI CODE BAGIAN INI
```python
    for i in range(k, len(b)):
        if a[i, k] != 0:
            a[[k, i]] = a[[i, k]]
            b[[k, i]] = b[[i, k]]
            break
```

Misalnya kalau matriks `a` kita ubah baris 1 kolom 1 nya jadi 0

```python
a = np.array([[2, 2, 1],
             [1, -1, 1],
             [3, 2, -2]], dtype=float)

b = np.array([3,4,2], dtype=float)
```

Matriks augmentasi kita akan menjadi seperti ini

$$
\left[
\begin{array}{ccc|c}
0 & 2 & 1 & 3 \\
1 & -1 & 1 & 4 \\
3 & 2 & -2 & 2
\end{array}
\right]
$$

Maka nanti pivot pertama ```a[0,0] = 0``` jadi code dibawah ini akan dijalankan
```python
    for i in range(k, len(b)):
        if a[i, k] != 0:
            a[[k, i]] = a[[i, k]]
            b[[k, i]] = b[[i, k]]
            break
```

` for i in range(k, 3): `

`k` = 0, `len(b)` = 3

i = 0, 1, 2

---

**Iterasi Pertama** → `i = 0, k = 0`

Cek kondisi `if a[0,0] != 0` 

Tapi `a[0,0] = 0` berarti kondisi `if a[i, k] != 0:` tidak terpenuhi

Maka code swap baris ini tidak dijalankan karena kita mau swap sama baris yang nilai pivotnya bukan 0 supaya bisa melanjutkan algoritma eliminasi gauss
```python
            a[[k, i]] = a[[i, k]]
            b[[k, i]] = b[[i, k]]
            break
```

---
**Iterasi Kedua** → `i = 1, k = 0`

Cek kondisi `if a[1,0] != 0`

`a[1,0] = 1` berarti kondisi `if a[i, k] != 0:` terpenuhi

Jalankan code swap baris

```python
            a[[k, i]] = a[[i, k]]
            b[[k, i]] = b[[i, k]]
            break
```

```
            a[[0,1]] = a[[1,0]] 
```

note: `[[ ]]` itu kayak `a[...]` untuk indexing dan `[1,0]` adalah list indexnya

`a[[1,0]]` artinya ambil 2 baris array berdasarkan list index, jadi pertama ambil baris ke-1 dulu (index 0), lalu ambil baris ke-0 

masukkan ke `a[[0,1]] = ... `

JADI
- baris 0 diisi baris 1
- baris 1 diisi baris 0

artinya baris ke-1 (index 0) ditukar dengan baris ke-2 (index 1)

---

TAPI itu baru matriks `a` nya, kita perlu tukar juga baris matriks `b`

Tadi kan `b = [3, 4, 2]`

` b[[0,1]] = b[[1,0]] `

`b[[1,0]]` index 1 = 4, index 0 = 3 hasilnya list `[4, 3]`

Jadi

`b[[0,1]] = [4, 3]`

Maka

`b[0] = 4`

`b[1] = 3`

---

**HASIL SETELAH SWAP**

$$
\left[
\begin{array}{ccc|c}
1 & -1 & 1 & 4 \\
0 & 2 & 1 & 3 \\
3 & 2 & -2 & 2
\end{array}
\right]
$$


Sekarang pivot sudah tidak nol, sehingga eliminasi Gauss bisa dilanjutkan.

---

### CATATAN #4
### Ubah pivot jadi 1

Matriks awal kita 
$$
\left[
\begin{array}{ccc|c}
2 & 2 & 1 & 3 \\
1 & -1 & 1 & 4 \\
3 & 2 & -2 & 2
\end{array}
\right]
$$


``` python
for k in range(n):
    ...
  
    a[k,:] /= pivot # a[k,:] = a[k,:] / pivot
    b[k] /= pivot   # b[k] = b[k] / pivot
    
    ...
    
```


- `n` = 3
- `k` = 0 → kolom pertama
- `k` = 1 → kolom kedua
- `k` = 2 → kolom ketiga

---

**Iterasi Pertama** → `k = 0, pivot = 2`

**Matriks a**

`a[k,:]` = `a[0,:]` = `[2, 2, 1]`

Semua elemen di baris 0 dibagi 2

`[2, 2, 1] = [2, 2, 1] / 2`

Hasilnya: [2, 2, 1] / 2 = [1, 1, 0.5]

**Matriks b**

`b[k]` = `b[0]` = `b[2]`

Elemen di baris 0 juga dibagi 2

`3 = 3 / 2`

Hasilnya: 3 / 2 = 1.5


---

### CATATAN #5 
### Ubah elemen di bawah pivot jadi 0

Matriks setelah langkah di atas
$$
\left[
\begin{array}{ccc|c}
1 & 1 & 0.5 & 1.5 \\
1 & -1 & 1 & 4 \\
3 & 2 & -2 & 2
\end{array}
\right]
$$


``` python
for k in range(n):
    ...
    
    for i in range(k+1, n):
        factor = a[i, k] 
        a[i, k:] -= factor * a[k, k:]
        b[i] -= factor * b[k]

    ...
```
`n` = 3

---

**Iterasi Pertama** → `k` = 0 
`for i in range(0+1, 3):` 
Berarti `i` = 1, 2

- Saat `i` = 1

``` python
        factor = a[i, k] 
```
```
        factor = a[1,0] = 1 
```
``` python
        a[i, k:] = a[i, k:] - factor * a[k, k:] 
```
```    
        a[1, 0:] = a[1, 0:] - 1 * a[0, 0:] 
```
``` python
        b[i] -= factor * b[k] 
```
```
        b[1] -= 1 * b[0]
```

---

Contoh perhitungan manual code yang ini (iterasi pertama)
``` python
        a[i, k:] = a[i, k:] - factor * a[k, k:] # matriks a
```
```    
        a[1, 0:] = a[1, 0:] - 1 * a[0, 0:] 
```
Note:
- baris di index 1, kolom dari 0 sampai akhir
- `a[1, 0:]` = [1  -1   1]   
- baris di index 0, kolom dari 0 sampai akhir  
- `a[0, 0:]` = [1   1   0.5] 

H21(1) [a]

```
[1  -1   1] = [1  -1     1] - 1 * [1   1   0.5] 

[1  -1   1] = [1  -1     1] - [1   1   0.5] 

[1  -1   1] = [0  -2   0.5] 
```
---
Contoh perhitungan manual code yang ini (iterasi pertama)

``` python
        b[i] = b[i] - factor * b[k] # matriks b
```
```
        b[1] = b[1] - 1 * b[0]
```

Note:
- `b[0]` = 1.5
- `b[1]` = 4


H21(1) [b]
```
4 = 4 - 1 * 1.5 

4 = 4 - 1.5

4 = 2.5
```

---

Matriks setelah iterasi 1
$$
\left[
\begin{array}{ccc|c}
1 & 1 & 0.5 & 1.5 \\
0 & -2 & 0.5 & 2.5 \\
3 & 2 & -2 & 2
\end{array}
\right]
$$

---


**Iterasi Kedua** → `k` = 1
`for i in range(1+1, 3):`
Berarti `i` = 2

**Iterasi Ketiga** → `k` = 2
`for i in range(2+1, 3):`
Berarti tidak ada


### CATATAN #6
### Setelah eliminasi gauss selesai (matriks udah berbentuk segitiga atas)

Matriks setelah Eliminasi Gauss

$$
\left[
\begin{array}{ccc|c}
1 & 1 & 0.5 & 1.5 \\
0 & 1 & -0.25 & -1.25 \\
0 & 0 & 1 & 1
\end{array}
\right]
$$

### Sekarang cari nilai x 

``` python
x = np.zeros(n)
```
```
x = [0, 0, 0]
```
``` python
x[-1] = b[-1] / a[-1, -1] # ambil variabel terakhir
```
```
x[2] = b[2] / a[2,2]

x[2] = 1 / 1 = 1
```

**Sebelum `x = [0, 0, 0]`**

**Sesudah `x = [0, 0, 2]`**

---

``` python
for i in range(n-2, -1, -1):
    sum_val = b[i] - np.dot(a[i, i+1:], x[i+1:])
    x[i] = sum_val / a[i, i]

```

**i = 1**
```
for i in range(1, -1, -1): # i = 1, 0
    
    # sum_val = -1.25 - np.dot(-0.25, 1)
    # sum_val = -1.25 + 0.25 
    # sum_val = -1
    sum_val= b[1] - np.dot(a[1, 2:], x[2:])

    # x[1] = -1 / 1
    # x[1] = -1
    x[1] = sum_val / a[1, 1]

```

Note
- baris ke-1, kolom mulai dari 2 sampai akhir (matriks a)
- a[1, 2:] = [-0.25]
- elemen dari index 2 sampai akhir (array x)
- x[2:] = [1]
- b[1] = -1.25
- a[1,1] = 1

**Sebelum `x = [0, 0, 2]`**

**Sesudah `x = [0, -1, 2]`**

---


``` python
for i in range(n-2, -1, -1):
    sum_val = b[i] - np.dot(a[i, i+1:], x[i+1:])
    x[i] = sum_val / a[i, i]

```

**i = 0**
```
for i in range(1, -1, -1): # i = 1, 0
    
    # sum_val = 1.5 - np.dot([1, 0.5], [-1, 1])
    # sum_val = 1.5 - (-0.5)
    # sum_val = 2
    sum_val = b[0] - np.dot(a[0, 1:], x[1:])

    # x[0] = 2 / 1
    # x[0] = 2
    x[0] = sum_val / a[0, 0]

```

Note
- baris ke-0, kolom mulai dari 1 sampai akhir (matriks a)
- a[0, 1:] = [1, 0.5]
- elemen dari index 1 sampai akhir (array x)
- x[1:] = [-1, 1]
- b[0] = 1.5
- a[0,0] = 1

**Sebelum `x = [0, -1, 1]`**

**Sesudah `x = [2, -1, 1]`**
